packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -416,6 +416,11 @@ class ConnectionStateHandler implements IConnectionStateHandler {
 	private applyForConnectedState(
 		source: "removeMemberEvent" | "addMemberEvent" | "timeout" | "containerSaved",
 	) {
+		if (this.connection === undefined) {
+			// we may arrive here due to ops processed from delta storage without any websocket connection
+			this.handler.logger.sendTelemetryEvent({ eventName: "applyWhileNotConnected", source });
```

Comment:
I'm not sure it's the right fix for the problem you described, or rather - this code might need more complete fix for the problem you describe (and likely we are missing UTs).
For example, I'd expect this.waitEvent?.end() to happen on disconnect with some new source type.

Reply:
This change may be wrong. I added this because the shouldWaitForJoinSignal() call below would trip an assert when there was no connection.

From testing I saw this happen from "removeMemberEvent" and "containerSaved". "removeMemberEvent" makes sense because when replaying saved ops from a stash blob, we may replay the previous client's join and leave messages. For "containerSaved", my only guess is that we get ops from storage but have no websocket connection, and these ops include all of our stashed ops, causing runtime to emit "saved".